Adapted for PHP 5.3.3 from

From 52b93f0cfd3cba7ff98cc5198df6ca4f23865f80 Mon Sep 17 00:00:00 2001
From: Stanislav Malyshev <stas@php.net>
Date: Sun, 5 Apr 2015 16:01:24 -0700
Subject: [PATCH] Fixed bug #69353 (Missing null byte checks for paths in
 various PHP extensions)

diff -up php-5.3.3/ext/dom/document.c.bug69353 php-5.3.3/ext/dom/document.c
--- php-5.3.3/ext/dom/document.c.bug69353	2010-04-02 22:08:15.000000000 +0200
+++ php-5.3.3/ext/dom/document.c	2015-06-10 09:22:30.589778441 +0200
@@ -1573,6 +1573,9 @@ static xmlDocPtr dom_document_parser(zva
 	xmlInitParser();
 
 	if (mode == DOM_LOAD_FILE) {
+		if (strlen(source) != source_len) {
+			return NULL;
+		}
 		char *file_dest = _dom_get_valid_file_path(source, resolved_path, MAXPATHLEN  TSRMLS_CC);
 		if (file_dest) {
 			ctxt = xmlCreateFileParserCtxt(file_dest);
@@ -1754,7 +1757,7 @@ PHP_FUNCTION(dom_document_save)
 		return;
 	}
 
-	if (file_len == 0) {
+	if (file_len == 0 || strlen(file) != file_len) {
 		php_error_docref(NULL TSRMLS_CC, E_WARNING, "Invalid Filename");
 		RETURN_FALSE;
 	}
@@ -1992,6 +1995,10 @@ static void _dom_document_schema_validat
 
 	switch (type) {
 	case DOM_LOAD_FILE:
+		if (strlen(source) != source_len) {
+			php_error_docref(NULL TSRMLS_CC, E_WARNING, "Invalid Schema file source");
+			RETURN_FALSE;
+		}
 		valid_file = _dom_get_valid_file_path(source, resolved_path, MAXPATHLEN  TSRMLS_CC);
 		if (!valid_file) {
 			php_error_docref(NULL TSRMLS_CC, E_WARNING, "Invalid Schema file source");
@@ -2081,6 +2088,10 @@ static void _dom_document_relaxNG_valida
 
 	switch (type) {
 	case DOM_LOAD_FILE:
+		if (strlen(source) != source_len) {
+			php_error_docref(NULL TSRMLS_CC, E_WARNING, "Invalid Schema file source");
+			RETURN_FALSE;
+		}
 		valid_file = _dom_get_valid_file_path(source, resolved_path, MAXPATHLEN  TSRMLS_CC);
 		if (!valid_file) {
 			php_error_docref(NULL TSRMLS_CC, E_WARNING, "Invalid RelaxNG file source");
@@ -2170,6 +2181,10 @@ static void dom_load_html(INTERNAL_FUNCT
 	}
 
 	if (mode == DOM_LOAD_FILE) {
+		if (strlen(source) != source_len) {
+			php_error_docref(NULL TSRMLS_CC, E_WARNING, "Invalid file source");
+			RETURN_FALSE;
+		}
 		ctxt = htmlCreateFileParserCtxt(source, NULL);
 	} else {
 		source_len = xmlStrlen(source);
@@ -2258,7 +2273,7 @@ PHP_FUNCTION(dom_document_save_html_file
 		return;
 	}
 
-	if (file_len == 0) {
+	if (file_len == 0 || strlen(file) != file_len) {
 		php_error_docref(NULL TSRMLS_CC, E_WARNING, "Invalid Filename");
 		RETURN_FALSE;
 	}
diff -up php-5.3.3/ext/fileinfo/fileinfo.c.bug69353 php-5.3.3/ext/fileinfo/fileinfo.c
--- php-5.3.3/ext/fileinfo/fileinfo.c.bug69353	2015-06-05 13:15:17.175945656 +0200
+++ php-5.3.3/ext/fileinfo/fileinfo.c	2015-06-05 13:15:25.081006435 +0200
@@ -485,6 +485,11 @@ static void _php_finfo_get_type(INTERNAL
 				RETVAL_FALSE;
 				goto clean;
 			}
+			if (strlen(buffer) != buffer_len) {
+				php_error_docref(NULL TSRMLS_CC, E_WARNING, "Invalid path");
+				RETVAL_FALSE;
+				goto clean;
+			}
 
 			wrap = php_stream_locate_url_wrapper(buffer, &tmp2, 0 TSRMLS_CC);
 
diff -up php-5.3.3/ext/gd/gd.c.bug69353 php-5.3.3/ext/gd/gd.c
--- php-5.3.3/ext/gd/gd.c.bug69353	2015-06-10 09:06:53.346558500 +0200
+++ php-5.3.3/ext/gd/gd.c	2015-06-10 09:18:19.187841895 +0200
@@ -1469,6 +1469,9 @@ PHP_FUNCTION(imageloadfont)
 	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s", &file, &file_name) == FAILURE) {
 		return;
 	}
+	if (strlen(file) != file_name) {
+		return;
+	}
 
 	stream = php_stream_open_wrapper(file, "rb", ENFORCE_SAFE_MODE | IGNORE_PATH | IGNORE_URL_WIN | REPORT_ERRORS, NULL);
 	if (stream == NULL) {
@@ -2421,6 +2424,9 @@ static void _php_image_create_from(INTER
 			return;
 		}
 	}
+	if (strlen(file) != file_len) {
+	    return;
+	}
 
 	stream = php_stream_open_wrapper(file, "rb", ENFORCE_SAFE_MODE|REPORT_ERRORS|IGNORE_PATH|IGNORE_URL_WIN, NULL);
 	if (stream == NULL)	{
@@ -4023,6 +4029,9 @@ PHP_FUNCTION(imagepsloadfont)
 	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s", &file, &file_len) == FAILURE) {
 		return;
 	}
+	if (strlen(file) != file_len) {
+		return;
+	}
 
 #ifdef PHP_WIN32
 	if (VCWD_STAT(file, &st) < 0) {
@@ -4130,6 +4139,9 @@ PHP_FUNCTION(imagepsencodefont)
 	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "rs", &fnt, &enc, &enc_len) == FAILURE) {
 		return;
 	}
+	if (strlen(enc) != enc_len) {
+		return;
+	}
 
 	ZEND_FETCH_RESOURCE(f_ind, int *, &fnt, -1, "Type 1 font", le_ps_font);
 
diff -up php-5.3.3/ext/hash/hash.c.bug69353 php-5.3.3/ext/hash/hash.c
--- php-5.3.3/ext/hash/hash.c.bug69353	2015-06-05 13:26:23.241066775 +0200
+++ php-5.3.3/ext/hash/hash.c	2015-06-05 13:26:35.415160377 +0200
@@ -136,6 +136,10 @@ static void php_hash_do_hash(INTERNAL_FU
 		RETURN_FALSE;
 	}
 	if (isfilename) {
+ 		if (strlen(data) != data_len) {
+			php_error_docref(NULL TSRMLS_CC, E_WARNING, "Invalid path");
+ 			RETURN_FALSE;
+ 		}
 		stream = php_stream_open_wrapper_ex(data, "rb", REPORT_ERRORS | ENFORCE_SAFE_MODE, NULL, DEFAULT_CONTEXT);
 		if (!stream) {
 			/* Stream will report errors opening file */
@@ -214,6 +218,10 @@ static void php_hash_do_hash_hmac(INTERN
 		RETURN_FALSE;
 	}
 	if (isfilename) {
+		if (strlen(data) != data_len) {
+			php_error_docref(NULL TSRMLS_CC, E_WARNING, "Invalid path");
+			RETURN_FALSE;
+		}
 		stream = php_stream_open_wrapper_ex(data, "rb", REPORT_ERRORS | ENFORCE_SAFE_MODE, NULL, DEFAULT_CONTEXT);
 		if (!stream) {
 			/* Stream will report errors opening file */
@@ -444,6 +452,9 @@ PHP_FUNCTION(hash_update_file)
 	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "rs|r", &zhash, &filename, &filename_len, &zcontext) == FAILURE) {
 		return;
 	}
+	if (strlen(filename) != filename_len) {
+		return;
+	}
 
 	ZEND_FETCH_RESOURCE(hash, php_hash_data*, &zhash, -1, PHP_HASH_RESNAME, php_hash_le_hash);
 	context = php_stream_context_from_zval(zcontext, 0);
diff -up php-5.3.3/ext/pgsql/pgsql.c.bug69353 php-5.3.3/ext/pgsql/pgsql.c
--- php-5.3.3/ext/pgsql/pgsql.c.bug69353	2015-06-05 13:28:47.897178979 +0200
+++ php-5.3.3/ext/pgsql/pgsql.c	2015-06-05 13:28:51.870209526 +0200
@@ -2868,7 +2868,9 @@ PHP_FUNCTION(pg_trace)
 	if (zend_parse_parameters(argc TSRMLS_CC, "s|sr", &z_filename, &z_filename_len, &mode, &mode_len, &pgsql_link) == FAILURE) {
 		return;
 	}
-
+	if (strlen(z_filename) != z_filename_len) {
+		return;
+	}
 	if (argc < 3) {
 		CHECK_DEFAULT_LINK(id);
 	}		
diff -up php-5.3.3/ext/standard/streamsfuncs.c.bug69353 php-5.3.3/ext/standard/streamsfuncs.c
--- php-5.3.3/ext/standard/streamsfuncs.c.bug69353	2015-06-05 13:30:25.786931617 +0200
+++ php-5.3.3/ext/standard/streamsfuncs.c	2015-06-05 13:30:39.829039581 +0200
@@ -1473,6 +1473,9 @@ PHP_FUNCTION(stream_resolve_include_path
 	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s", &filename, &filename_len) == FAILURE) {
 		return;
 	}
+	if (strlen(filename) != filename_len) {
+		return;
+	}
 
 	resolved_path = zend_resolve_path(filename, filename_len TSRMLS_CC);
 
diff -up php-5.3.3/ext/xmlwriter/php_xmlwriter.c.bug69353 php-5.3.3/ext/xmlwriter/php_xmlwriter.c
--- php-5.3.3/ext/xmlwriter/php_xmlwriter.c.bug69353	2015-06-05 13:32:08.896724389 +0200
+++ php-5.3.3/ext/xmlwriter/php_xmlwriter.c	2015-06-05 13:32:22.661830223 +0200
@@ -1762,6 +1762,9 @@ static PHP_FUNCTION(xmlwriter_open_uri)
 	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s", &source, &source_len) == FAILURE) {
 		return;
 	}
+	if (strlen(source) != source_len) {
+		return;
+	}
 	
 #ifdef ZEND_ENGINE_2
 	if (this) {
diff -up php-5.3.3/ext/zlib/zlib.c.bug69353 php-5.3.3/ext/zlib/zlib.c
--- php-5.3.3/ext/zlib/zlib.c.bug69353	2015-06-05 13:33:41.050432921 +0200
+++ php-5.3.3/ext/zlib/zlib.c	2015-06-05 13:34:14.541690425 +0200
@@ -401,6 +401,9 @@ static PHP_FUNCTION(gzfile)
 	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s|l", &filename, &filename_len, &flags) == FAILURE) {
 		return;
 	}
+	if (strlen(filename) != filename_len) {
+		return;
+	}
 
 	use_include_path = flags ? USE_PATH : 0;
 
@@ -444,6 +447,9 @@ static PHP_FUNCTION(gzopen)
 	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "ss|l", &filename, &filename_len, &mode, &mode_len, &flags) == FAILURE) {
 		return;
 	}
+	if (strlen(filename) != filename_len) {
+		return;
+	}
 
 	use_include_path = flags ? USE_PATH : 0;
 
@@ -473,6 +479,9 @@ static PHP_FUNCTION(readgzfile)
 	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s|l", &filename, &filename_len, &flags) == FAILURE) {
 		return;
 	}
+	if (strlen(filename) != filename_len) {
+		return;
+	}
 
 	use_include_path = flags ? USE_PATH : 0;
 
