diff -up php-5.3.3/ext/imap/php_imap.c.859371 php-5.3.3/ext/imap/php_imap.c
--- php-5.3.3/ext/imap/php_imap.c.859371	2012-10-09 10:25:27.695781670 +0200
+++ php-5.3.3/ext/imap/php_imap.c	2012-10-09 10:25:52.887859821 +0200
@@ -1190,7 +1190,7 @@ static void php_imap_do_open(INTERNAL_FU
 							if (zend_hash_index_find(Z_ARRVAL_PP(disabled_auth_method), i, (void **) &z_auth_method) == SUCCESS) {
 								if (Z_TYPE_PP(z_auth_method) == IS_STRING) {
 									if (Z_STRLEN_PP(z_auth_method) > 1) {
-										mail_parameters (NIL, DISABLE_AUTHENTICATOR, (void *)Z_STRVAL_PP(disabled_auth_method));
+										mail_parameters (NIL, DISABLE_AUTHENTICATOR, (void *)Z_STRVAL_PP(z_auth_method));
 									}
 								} else {
 									php_error_docref(NULL TSRMLS_CC, E_WARNING, "Invalid argument, expect string or array of strings");
